### PR TITLE
[SEDONA-235] Integrate S2, add ST_S2CellIDs

### PR DESCRIPTION
**Pull request by zongsizhang**
## Did you read the Contributor Guide?

- Yes, I have read [Contributor Rules](https://sedona.apache.org/latest-snapshot/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest-snapshot/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes, the URL of the associated JIRA ticket is [https://issues.apache.org/jira/browse/SEDONA-235](https://issues.apache.org/jira/browse/SEDONA-235). The PR name follows the format `[SEDONA-XXX] my subject`.

## What changes were proposed in this PR?

Add integration with Google S2 into Common https://s2geometry.io/
Add function to cover a geometry with s2Cells and return the corresponding cellIds.
Add spark core and sql functions to generate S2Cells from gemetry.

## How was this patch tested?

Add many unit tests in common, sql, flink, and python module

## Did this PR include necessary documentation updates?

- Yes, I am adding a new API. I am using the [current SNAPSHOT version number]

### CODE REVIEW

File: common/src/main/java/org/apache/sedona/common/utils/GeomUtils.java [hunk 1]
```diff
@@ -27,6 +27,8 @@
 
 import java.nio.ByteOrder;
 import java.util.*;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
```
Comment (Kimahriman) [inline common/src/main/java/org/apache/sedona/common/utils/GeomUtils.java:31]: Doesn't look like these are used

File: common/src/main/java/org/apache/sedona/common/Functions.java [hunk 2]
```diff
@@ -543,4 +542,25 @@ public static Geometry split(Geometry input, Geometry blade) {
...
+     * @param input Geometry
+     * @param level integer, minimum level of cells covering the geometry
+     * @return List of coordinates
+     */
+    public static Long[] s2CellIDs(Geometry input, int level) {
+        HashSet<S2CellId> cellIds = new HashSet<>();
+        List<Geometry> geoms = GeomUtils.extractGeometryCollection(input);
+        for (Geometry geom : geoms) {
+            try {
+                cellIds.addAll(S2Utils.s2RegionToCellIDs(S2Utils.toS2Region(geom), 1, level, Integer.MAX_VALUE - 1));
```
Comment (jiayuasu) [inline common/src/main/java/org/apache/sedona/common/Functions.java:558]: More case handlers are needed. Currently, Sedona supports 7 most common geometry types

instanceOf Point
instanceOf MuliPoint
instanceOf Polygon
instanceOf MultiPolygon
instanceOf LineString
instanceOf MultiLineString
instanceOf GeometryCollection

File: common/src/test/java/org/apache/sedona/common/FunctionsTest.java [hunk 3]
```diff
@@ -238,4 +235,98 @@ public void splitHeterogeneousGeometryCollection() {
 
         assertNull(actualResult);
     }
+
+    @Test
```
Comment (jiayuasu) [inline common/src/test/java/org/apache/sedona/common/FunctionsTest.java:239]: Please add more tests for different types (7 types in total)

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala [hunk 4]
```diff
@@ -1055,3 +1055,11 @@ case class ST_Split(inputExpressions: Seq[Expression])
     copy(inputExpressions = newChildren)
   }
 }
+
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala:1058]: Please also add this SQL function to Sedona Flink module

File: docs/api/sql/Function.md [hunk 5]
```diff
@@ -1536,3 +1536,20 @@ SELECT ST_ZMin(ST_GeomFromText('LINESTRING(1 3 4, 5 6 7)'))
...
+
+## ST_S2CellIDs
+
+Introduction: Cover the geometry with Google S2 Cells, return the corresponding cell IDs with the given level.
+The level indicates the [size of cells](https://s2geometry.io/resources/s2cell_statistics.html). With a bigger level,
+the cells will be smaller, the coverage will be more accurate, but the result size will be exponentially increasing.
+
+Format: `ST_S2CellIDs(geom: geometry, level: Int)`
+
+Since: `v1.3.2`
```
Comment (jiayuasu) [inline docs/api/sql/Function.md:1548]: Please use version 1.4.0

**Review by jiayuasu — CHANGES_REQUESTED** [review]

File: common/src/main/java/org/apache/sedona/common/Functions.java (hunk 2 above)
Comment (zongsizhang) [reply common/src/main/java/org/apache/sedona/common/Functions.java:558]: they are all supported already. MultiPoint, MultiPolygon, MultiLineString are GeometryCollection, after extraction they become points, polygons and linestrings.

**Comment by zongsizhang** [comment]
@jiayuasu do you know where is this error from?
https://github.com/apache/sedona/actions/runs/4157076083/jobs/7191302226#step:18:218

File: common/src/test/java/org/apache/sedona/common/FunctionsTest.java (hunk 3 above)
Comment (zongsizhang) [reply common/src/test/java/org/apache/sedona/common/FunctionsTest.java:245]: added test case for all 7 types

File: common/src/main/java/org/apache/sedona/common/utils/GeomUtils.java (hunk 1 above)
Comment (zongsizhang) [reply common/src/main/java/org/apache/sedona/common/utils/GeomUtils.java]: removed

File: sql/pom.xml [hunk 6]
```diff
@@ -123,6 +123,11 @@
             <groupId>org.scalatest</groupId>
             <artifactId>scalatest_${scala.compat.version}</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.google.geometry</groupId>
+            <artifactId>s2-geometry</artifactId>
+            <version>2.0.0</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (zongsizhang) [reply sql/pom.xml]: thanks have removed it

File: common/pom.xml [hunk 7]
```diff
@@ -57,6 +57,11 @@
             <groupId>org.wololo</groupId>
             <artifactId>jts2geojson</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.google.geometry</groupId>
+            <artifactId>s2-geometry</artifactId>
+            <version>2.0.0</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (zongsizhang) [reply common/pom.xml]: thanks, change committed, please help review

File: common/src/test/java/org/apache/sedona/common/FunctionsTest.java [hunk 8]
```diff
@@ -238,4 +235,153 @@ public void splitHeterogeneousGeometryCollection() {
...
+    public void getGoogleS2CellIDsMultiLineString() {
+        // polygon with holes
+        MultiLineString target = GEOMETRY_FACTORY.createMultiLineString(
+                new LineString[] {
+                        GEOMETRY_FACTORY.createLineString(coordArray(0.1, 0.1, 0.2, 0.1, 0.3, 0.4, 0.5, 0.9)),
+                        GEOMETRY_FACTORY.createLineString(coordArray(0.5, 0.1, 0.1, 0.5, 0.3, 0.1))
+                }
+        );
+        Geometry mbr = target.getEnvelope();
+        Point outsidePoint = GEOMETRY_FACTORY.createPoint(new Coordinate(0.3, 0.7));
```
Comment (jiayuasu) [inline common/src/test/java/org/apache/sedona/common/FunctionsTest.java:318]: This variable is not used.

File: sql/src/main/scala/org/apache/sedona/sql/UDF/Catalog.scala [hunk 9]
```diff
@@ -167,7 +167,8 @@ object Catalog {
     function[RS_HTML](),
     function[RS_Array](),
     function[RS_Normalize](),
-    function[RS_Append]()
+    function[RS_Append](),
+    function[ST_S2CellIDs]()
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/sedona/sql/UDF/Catalog.scala:171]: This should be put together with other ST functions.

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/st_functions.scala [hunk 10]
```diff
@@ -271,4 +271,7 @@ object st_functions extends DataFrameAPI {
 
   def ST_ZMin(geometry: Column): Column = wrapExpression[ST_ZMin](geometry)
   def ST_ZMin(geometry: String): Column = wrapExpression[ST_ZMin](geometry)
+
+  def ST_S2CellIDs(geometry: Column, level: Int): Column = wrapExpression[ST_S2CellIDs](geometry, level)
+  def ST_S2CellIDs(geometry: String, level: Int): Column = wrapExpression[ST_S2CellIDs](geometry, level)
```
Comment (jiayuasu) [inline sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/st_functions.scala:276]: (1) Please add simple tests for this DataFrame style API.
(2) Please add this DataFrame style API to Sedona Python as well.

**Review by jiayuasu — CHANGES_REQUESTED** [review]

**Comment by jiayuasu** [comment]
> @jiayuasu do you know where is this error from? [apache/sedona/actions/runs/4157076083/jobs/7191302226#step:18:218](https://github.com/apache/sedona/actions/runs/4157076083/jobs/7191302226#step:18:218)

@zongsizhang Sedona R CI is currently failing due to some unknown reason on the dependencies used in Sedona R module. We need our R maintainer to investigate this issue. Please ignore it for now.

File: common/src/test/java/org/apache/sedona/common/FunctionsTest.java (hunk 8 above)
Comment (zongsizhang) [reply common/src/test/java/org/apache/sedona/common/FunctionsTest.java:318]: outsidePoint? it's used at row 321
```Long outsideCell = Functions.s2CellIDs(outsidePoint, 10)[0];```

File: docs/api/sql/Function.md [hunk 11]
```diff
@@ -1536,3 +1536,20 @@ SELECT ST_ZMin(ST_GeomFromText('LINESTRING(1 3 4, 5 6 7)'))
 ```
 
 Output: `4.0`
+
+## ST_S2CellIDs
```
[the remark this answers is not in the document]
Comment (zongsizhang) [reply docs/api/sql/Function.md]: changed, please check

File: sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/st_functions.scala (hunk 10 above)
Comment (zongsizhang) [reply sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/st_functions.scala]: done, please check

File: docs/api/sql/Function.md [hunk 12]
```diff
@@ -1101,6 +1101,30 @@ Result:
 +---------------------------------------------------------------+
 ```
 
+## ST_S2CellIDs
```
Comment (jiayuasu) [inline docs/api/sql/Function.md:1104]: Please also add this function to Sedona Flink doc.

**Review by jiayuasu — CHANGES_REQUESTED** [review]

**Review by jiayuasu — APPROVED** [review]

**Pull request merged**